pkg/core/native/oracle.go

```diff
@@ -0,0 +1,432 @@
+	o.AddMethod(md, desc, true)
+
+	desc = newDescriptor("setOracleNodes", smartcontract.VoidType)
+	md = newMethodAndPrice(o.setOracleNodes, 0, smartcontract.AllowModifyStates)
+	o.AddMethod(md, desc, false)
+
+	desc = newDescriptor("verify", smartcontract.BoolType)
+	md = newMethodAndPrice(o.verify, 100_0000, smartcontract.NoneFlag)
+	o.AddMethod(md, desc, false)
+
+	pp := chainOnPersist(postPersistBase, o.PostPersist)
+	desc = newDescriptor("postPersist", smartcontract.VoidType)
+	md = newMethodAndPrice(getOnPersistWrapper(pp), 0, smartcontract.AllowModifyStates)
+	o.AddMethod(md, desc, false)
+
+	return o
+}
+
+// PostPersist represents `postPersist` method.
+func (o *Oracle) PostPersist(ic *interop.Context) error {
+	var nodes keys.PublicKeys
+	var reward []big.Int
+	single := new(big.Int).SetUint64(oracleRequestPrice)
+	for _, tx := range ic.Block.Transactions {
+		resp := getResponse(tx)
+		if resp == nil {
+			continue
+		}
+		reqKey := makeRequestKey(resp.ID)
+		req := new(OracleRequest)
+		if err := o.getSerializableFromDAO(ic.DAO, reqKey, req); err != nil {
+			return err
+		}
+		if err := ic.DAO.DeleteStorageItem(o.ContractID, reqKey); err != nil {
+			return err
+		}
+
+		idKey := makeIDListKey(req.URL)
+		idList := new(IDList)
+		if err := o.getSerializableFromDAO(ic.DAO, idKey, idList); err != nil {
+			return err
+		}
+		if !idList.Remove(resp.ID) {
+			return errors.New("response ID wasn't found")
+		}
+
+		var err error
+		if len(*idList) == 0 {
+			err = ic.DAO.DeleteStorageItem(o.ContractID, idKey)
+		} else {
+			si := &state.StorageItem{Value: idList.Bytes()}
+			err = ic.DAO.PutStorageItem(o.ContractID, idKey, si)
+		}
+		if err != nil {
+			return err
+		}
+
+		if nodes == nil {
+			nodes = o.GetOracleNodes()
+			reward = make([]big.Int, len(nodes))
+		}
+
+		if len(reward) > 0 {
+			index := resp.ID % uint64(len(nodes))
+			reward[index].Add(&reward[index], single)
+		}
+	}
+	for i := range reward {
+		o.GAS.mint(ic, nodes[i].GetScriptHash(), &reward[i])
+	}
+	return nil
+}
+
+// Metadata returns contract metadata.
+func (o *Oracle) Metadata() *interop.ContractMD {
+	return &o.ContractMD
+}
+
+// Initialize initializes Oracle contract.
+func (o *Oracle) Initialize(ic *interop.Context) error {
+	si := &state.StorageItem{Value: NodeList{}.Bytes()}
+	if err := ic.DAO.PutStorageItem(o.ContractID, prefixNodeList, si); err != nil {
+		return err
+	}
+	o.nodes.Store(keys.PublicKeys(nil))
+	si = &state.StorageItem{Value: make([]byte, 8)} // uint64(0) LE
+	return ic.DAO.PutStorageItem(o.ContractID, prefixRequestID, si)
+}
+
+func getResponse(tx *transaction.Transaction) *transaction.OracleResponse {
+	for i := range tx.Attributes {
+		if tx.Attributes[i].Type == transaction.OracleResponseT {
+			return tx.Attributes[i].Value.(*transaction.OracleResponse)
+		}
+	}
+	return nil
+}
+
+func (o *Oracle) finish(ic *interop.Context, _ []stackitem.Item) stackitem.Item {
+	err := o.FinishInternal(ic)
+	if err != nil {
+		panic(err)
```

Comment:
This is worth adding a test for, callback errors should be forwarded into transaction errors.

Reply:
Added a test for this case.